### PR TITLE
cpu: eltwise: Fix crash in eltwise_pow on Windows

### PR DESCRIPTION
**Pull request by markdryan**
The primitive is crashing as the code it generates to ensure that the
stack is 16 byte aligned actually ensures that it is always 8 byte
aligned.  We should be adding 32 + the potential alignment offset
instead of 40 + the alignment offset, as described in the comments
in the code.

Signed-off-by: Mark Ryan <mark.d.ryan@intel.com>

# Description

@densamoilov can you check this?  The issue seems to have been introduced by [#1cc4fd7291cdfdba43a3cf0fa71265c02ac47d93](https://github.com/oneapi-src/oneDNN/commit/1cc4fd7291cdfdba43a3cf0fa71265c02ac47d93).  Prior to this commit the code arranged for the stack to always be 16 byte aligned, and after this commit it is always 8 byte aligned, as far as I can tell.  For example consider that the stack is 16 byte aligned at the start of this code

```
h->mov(h->rbx, h->rsp);
h->and_(h->rbx, 0xf);
h->add(h->rbx, 0x28);
h->sub(h->rsp, h->rbx);
```

line 2 will set rbx to 0, line 3 will set it to 40 (which is 8 byte aligned and not 16 byte aligned) and then line 4 will subtract 40 from rsp, making it 8 byte aligned, where previously, it had been 16 byte aligned.

# Checklist

## General

- [ ] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
  - I ran ctest and can see the eltwise tests were run
- [ ] Have you formatted the code using clang-format?

### Bug fixes

- [ ] Have you included information on how to reproduce the issue (either in a github issue or in this PR)?

I see the issue when cross compiling oneDNN for Windows using mingw64.  I have some custom code that just processes an array of floats with the eltwise_pow primitive.  The code crashes on Windows.  There's no crash if I build and run on Linux.  Applying this commit fixes the crash on Windows.

### CODE REVIEW

**Comment by densamoilov** [comment]
Hi @markdryan, my understanding is that a requirement to stack alignment prior to calling a function is to be 8 byte aligned because the call will push the return address on stack making the stack 16 byte aligned.

When making the changes I used [this](https://github.com/simon-whitehead/assembly-fun/blob/master/windows-x64/README.md#shadow-space) cheat sheet about 64-bit Windows ABI. I also checked [godbolt](https://godbolt.org/z/xdYY9n4Ef) where we can see that 40 bytes of stack is allocated prior to calling `powf`.

Do you have any reference documents that would back up the approach you are suggesting?

**Comment by markdryan** [comment]
> Hi @markdryan, my understanding is that a requirement to stack alignment prior to calling a function is to be 8 byte aligned because the call will push the return address on stack making the stack 16 byte aligned.
> 
> Do you have any reference documents that would back up the approach you are suggesting?

I came across [this](https://docs.microsoft.com/en-us/cpp/build/stack-usage?view=msvc-170#stack-allocation).  If you read through the paragraph you'll encounter this comment.

> The stack will always be maintained 16-byte aligned, except within the prolog **(for example, after the return address is pushed)**, and except where indicated in [Function Types](https://docs.microsoft.com/en-us/cpp/build/stack-usage?view=msvc-170#function-types) for a certain class of frame functions.

Which suggests that the stack should be 16 byte aligned before the call so it can be 8-byte aligned after the call.

Also in the [document you reference](https://github.com/simon-whitehead/assembly-fun/blob/master/windows-x64/README.md#stack-alignment) it says

> The simple act of calling a function misaligns the stack by placing an 8 byte return address on the stack when entering a function.

Also, implying that the stack should be 16 byte aligned before making the call and unaligned directly afterward.

There's also the fact that the current code crashes, at least for me, and that the old code prior to https://github.com/oneapi-src/oneDNN/commit/1cc4fd7291cdfdba43a3cf0fa71265c02ac47d93 explicitly arranged for the stack to be 16 byte aligned before the call.  https://github.com/oneapi-src/oneDNN/commit/1cc4fd7291cdfdba43a3cf0fa71265c02ac47d93 did two things.  It added the shadow space and it changed the alignment of the stack before the call.  Was the alignment change made to fix an issue?

**Comment by densamoilov** [comment]
We discussed the issue offline and came to the conclusion that we should preserve 16-byte alignment for the "shadow space".

**Review by densamoilov — APPROVED** [review]

File: src/cpu/x64/injectors/jit_uni_eltwise_injector.cpp [hunk 1]
```diff
@@ -1056,12 +1056,12 @@ void jit_uni_eltwise_injector_f32<isa, Wmm>::pow_compute_vector_fwd(
...
         // The 64-bit Windows ABI requires the caller to allocate 32 bytes of
-        // a so called "shadow space" for the callee. Since the return address
-        // is pushed on stack upon calling `powf` the `rsp` becomes 8-byte
-        // aligned. In order to allocate the shadow space and preserve the
-        // 16-byte alignment we have to allocate 40 bytes (32 bytes for the
-        // "shadow space" + 8 bytes to preserve 16-byte alignment).
-        h->add(h->rbx, 0x28);
+        // a so called "shadow space" for the callee.  In order to allocate the
+        // shadow space and preserve the 16-byte alignment we may actually need
+        // to allocate 40 bytes (32 bytes for the "shadow space" + 8 bytes to
```
[the remark this answers is not in the document]
Comment (markdryan) [reply src/cpu/x64/injectors/jit_uni_eltwise_injector.cpp:1061]: I guess it is still sort of relevant as at line 1065 we subtract either 32 bytes from rsp, if rsp is already aligned, or 40 bytes if it is not.  The comment might make a bit more sense if it came before the **and** instruction at line 1057, as it describes the combined behaviour of lines 1057, 1064 and 1065.

[the remark this answers is not in the document]
Comment (densamoilov) [reply src/cpu/x64/injectors/jit_uni_eltwise_injector.cpp:1061]: > The comment might make a bit more sense if it came before the and instruction at line 1057

Sounds good to me.

**Pull request merged**